.github/CONTRIBUTING.md

```diff
@@ -86,7 +121,43 @@ The communities behind these PHP frameworks/CMSes/polyfill libraries are strongl
 * Test case files should be placed in the same directory as the unit test file.
 
 
-## Running the Sniff Tests
+## Writing Sniff Tests
+
+Each sniff has its own remit and should function independently of other sniffs.
+
+With that in mind, sniffs are, and should be, tested in isolation.
+The tests for each sniff should only test the functionality of that specific sniff, independently of that the code used to test the sniff may also trigger errors from other sniffs.
+
+Tests for PHPCompatibility sniffs will always consist of sets of two files:
+1. A test case file (the `.inc` file) containing code snippets to run the sniff over.
+    Typically, a test case file should/could contain four closely related categories of code snippets:
+    1. "Not our target" - code which could confuse the sniff, but is not what the sniff is looking for.
+        For example, if a sniff is looking for a function call to a global PHP function, typical "not our target" code snippets would be a function call to a namespaced function with the same name as the global PHP function and method calls to methods with the same name as the global PHP function.
+    2. "OK/valid cross version" - code which the sniff is targetting, but which is not problematic (for the purposes of this particular sniff).
+        For example, if a sniff is looking for a new modifier keyword being allowed for OO properties, property declarations with modifiers _other_ than the one the sniff is looking for would fall into this category.
+    3. "Undetermined/ignored by the sniff" - code for which PHPCS is not typically suitable to determine whether the code is cross-version compatible or not.
+        For example, if a sniff is looking for a particular parameter type being passed to a function call, the sniff can flag this if the parameter is passed as a hard-coded value, like `10` or `false`, but the sniff cannot determine what the return type is of a function call, or find the type of a class constant declared elsewhere. In that case, the sniff should stay silent.
+    4. "The issue" - code which should be flagged by the sniff for having a PHP cross-version compatibility issue.
+        Please make sure the code snippets in this category use plenty of variation in whitespace, comment use and otherwise variation in how code can be written.
+        Sniffs should not expect to be run on "clean" codebases, so make sure to include code snippets written in the most horrible and most unreadable way you can imagine to make sure the sniff is properly tested.
```

Comment:
```suggestion
        Sniffs should not expect to be run on "clean" codebases, so make sure to include code snippets written in the most horrible and most unreadable way you can imagine to make sure the sniff is properly tested. Code might also be incomplete (user is still typing in their code editor and PHPCS runs as a plugin) or contain preexisting parsing errors, such as `Foo::`. It's best to put these at the end of the `.inc` file.
```
Added note about parsing errors and live coding, as that might come as a surprise to many.